accepted/2021/net6.0-tfms/net6.0-tfms.md

```diff
@@ -0,0 +1,329 @@
+Catalyst, but for the time being she's unblocked and can continue to develop her
+app.
+
+## Requirements
+
+### Goals
+
+* Support the definition of frameworks relevant for Xamarin development
+* Mac Catalyst is able to consume libraries built for Xamarin iOS, even though
+  the binary might not work. The developer should get a warning, but the tools
+  shouldn't block it.
+
+### Non-Goals
+
+* Re-design the TFM syntax
+
+## Stakeholders and Reviewers
+
+* MSBuild / SDK team
+* NuGet team
+* Project system team
+* Xamarin team
+
+## Design
+
+| TFM                | Compatible With                          |
+|--------------------|------------------------------------------|
+| net6.0             | (subsequent version of net5.0)           |
+| net6.0-windows     | (subsequent version of net5.0-windows)   |
+| net6.0-android     | xamarin.android                          |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-ios         | xamarin.ios                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-macos       | xamarin.mac                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-maccatalyst | xamarin.ios                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-tvos        | xamarin.tvos                             |
+|                    | (+everything else inherited from net6.0) |
+
+### Compatibility rules
+
+The [built-in compatibility rules][net5.0] of the `netX.Y-platform` TFMs make a
+TFM automatically compatible with:
+
+1. Itself
+2. Earlier versions of itself
+3. The corresponding `netX.Y` TFM
+4. Earlier versions of the corresponding `netX.Y` TFM
+
+The `net5.0` TFM has the additional twist that it's compatible with
+`netcoreapp3.1` (and earlier), `netstandard2.1` (and earlier) as well as .NET
+Framework (`net4x` and earlier, but under a warning). The .NET Framework
+compatibility is achieved as a *fallback*, that is, it's only used when nothing
+else applies.
+
+We don't want to model the Xamarin TFMs as a fallback but instead want them to
+apply **before** `netcoreapp` and `netstandard`. Usually the existence of the
+Xamarin TFM implementation is to bait a `netstandard` API and switch the
+platform specific implementation. Most of the Plugin type packages use this
+pattern and in fact the `netstandard` "implementation" is more like a reference
+assembly with just stubs that throw at runtime. `Xamarin.Essentials` is a good
+example of this. The same rational can be applied to `netcoreapp` where this
+would logically be the server implementation (or maybe a .NET Core 3.x desktop
+implementation).
+
+A `net6.0-maccatalyst` project referencing a NuGet package should behave as
+follows:
+
+* It should prefer `xamarin.ios` assets over `netcoreapp` and `netstandard`
+  assets.
+    - However, it should still prefer `net5.0`/`net6.0` assets over
```

Comment:
This seems wrong to me but it could be because I don't understand net5/6. Why would you ever want a less platform specific implementation over a more specific implementation? I would be very surprised to reference a library and find out it can't present UI because the net5/6 was chosen over ios version.

Reply:
`xamarin.ios` is the old Xamarin TFM. The assumption here is that if a package is updated, it should target iOS/Catalyst using the new TFMs (`net6.0-ios` and `net6.0-maccatalyst` respectively).

Reply:
OK, so the requirement is that if someone updates to net5/6 then they should update their iOS TFMs too? Logical but no obvious. Gonna have to spread that tip.

Reply:
> OK, so the requirement is that if someone updates to net5/6 then they should update their iOS TFMs too? Logical but no obvious.

My assumption, and I could totally be wrong here, is that we probably want support producing the old Xamarin TFM from the new tooling. So I'd think once you upgrade the tools to .NET 6, you're naturally already on the new TFMs anyway.

Reply:
I feel like this migration is going much slower than you think. I anticipate a lot of my apps will be mixed for years and I would prefer nugets to keep working. I'm thinking there will be a lot of nugets out there with ALL these TFMs. That said, as long as the tooling doesn't regress on the old TFMs, that old code will keep working.

Reply:
> @terrajobst I hand author all my nuspecs. 😅

![](https://media0.giphy.com/media/3o7527pa7qs9kCG78A/giphy.gif)

> I feel like this migration is going to be much slower than you think. I anticipate a lot of my apps will be mixed for years 

That's fair

> and I would prefer nugets to keep working.

Also fair; let me think about the compatibility relationship. While I'm all for making more stuff work I'm also very concerned with having rules that are simple enough to ambiguities and surprises. Let me chew on that.

Reply:
I tend to also think that if you're adding `net6.0-ios` support, it's not a stretch to expect you to add `net6.0-maccatalyst` as another target as well.  The tooling will not be super helpful in allowing your to construct a nupkg with a mix of old `Xamarin.iOS` and newer `net6.0-ios` TFM's to begin with unless you're producing artisanal, hand-crafted nupkg's like Frank, so I would expect many package authors to drop support for the old TFM at this point.  Even if you were going to mix `net6.0-ios` support with `Xamarin.iOS` support in the same package, it's not too crazy to throw `net6.0-maccatalyst` in there as well, right?

I think if you had all 3 TFM's (`net6.0-ios`, `net6.0-maccatalyst`, and `Xamarin.iOS`) in your nupkg, you could still support both old Xamarin and new .NET6 without issue.  

I think the only correction that might be needed here is to say `net6.0 assets are preferred' instead of net6.0 _and_ net5.0 since Xamarin was never supported on net5.0.  

The rule would be, make sure you have `net6.0-ios` and `net6.0-maccatalyst` in your packages if you want to support both iOS and macOS in a .NET 6 app, but you could still include `Xamarin.iOS` assets if you want to support apps targeting the old TFM.

Reply:
This is experimental (but maybe collection of hacks is a better description): https://github.com/xamarin/Xamarin.Legacy.Sdk

What this thing does is allow `dotnet build` running under .NET 6 to load the legacy Xamarin MSBuild targets in a multi-targeted project. So starting with something like a `net6.0-android;net6.0-ios` class library, you can add `monoandroid11.0;xamarin.ios10`.

The question is: how many packages will need to even use Xamarin.Legacy.Sdk? We've been thinking only components Xamarin maintains would need to do this. Things like AndroidX or Google Play Services.

Reply:
I’m pretty satisfied with the arguments here. It’s logical and @Redth seems to have the bases covered.

Reply:
I get that there are arguments for preferring platform specific over generational compatibility, but another perspective is why is a fallback, something that *might* work, being preferred over something that *will work*.

Reply:
> I get that there are arguments for preferring platform specific over generational compatibility, but another perspective is why is a fallback, something that _might_ work, being preferred over something that _will work_.

In this case the "something that will work" while technically correct from an API availability perspective, is not really true in practice since the `net5.0`  assets would not contain the platform specific 'switch' implementation that is required to be useful for these platforms.  So, yes, maybe things compile, but at runtime they will throw or otherwise be completely unusable.  In many cases in the Xamarin ecosystem, if there are `xamarinios` assets in a package, it is not just adding value on top of the `netstandard` assets, but rather it's a full bait and switch where the `netstandard` assets are really just considered a reference assembly with the public API contracts and stub implementations which throw at runtime.

My recommendations:
 - `net6.0-maccatalyst` prefers fallback path (in order) of: `net6.0`, `xamarinios`, `netstandard` 
 - `net6.0-ios` prefers the fallback path (in order) of `net6.0`, `xamarinios`, `netstandard`
 - The fallback paths explicitly exclude `net5.0` assets
 
Since Xamarin was never compatible in the net5 generation of TFM's and we're already seeing some packages (such as ReactiveUI) which ship assets for both `net5.0` and `xamarinios` today, it's important that we don't prefer `net5.0` assets over `xamarinios`.  If you have a `xamarinios` project today you'll get `xamarinios` assets instead of `net5.0` assets, and changing this for projects upgraded to `net6.0-ios` would break/change the expected behaviour.

Reply:
> * `net6.0-maccatalyst` prefers fallback path (in order) of: `net6.0`, `xamarinios`, `netstandard`

This makes sense. I have updated the spec accordingly.

> * `net6.0-ios` prefers the fallback path (in order) of `net6.0`, `xamarinios`, `netstandard`

That doesn't make sense to me. Why would you prefer `net6.0` over `xamarinios`? I thought we'd want to treat `xamarinios` as the previous version of `net6.0-ios`, which is basically how `net5.0` is treated to `netcoreapp`.

Reply:
I'm probably a bit pedantic here, but we should avoid using the term fallback here. 
Fallback is usually correlated with AssetTargetFallback, and this isn't that. Sounds like the ask is to have explicit compatibility between these frameworks.